scylla.go

```diff
@@ -484,53 +488,128 @@ func (p *scyllaConnPicker) GetCustomDialer() Dialer {
 		}
 	}
 
 	// We did not find an unallocated shard
 	// We will dial the non-shard-aware port
-	return nil
+	return 0, 0
+}
+
+// ShardDialer is like HostDialer but is shard-aware.
+// If the driver wants to connect to a specific shard, it will call DialShard,
+// otherwise it will call DialHost.
+type ShardDialer interface {
+	HostDialer
+
+	// DialShard establishes a connection to the specified shard ID out of nrShards.
+	// The returned connection must be directly usable for CQL protocol,
+	// specifically DialShard is responsible also for setting up the TLS session if needed.
+	DialShard(ctx context.Context, host *HostInfo, shardID, nrShards int) (*DialedHost, error)
 }
 
 // A dialer which dials a particular shard
-type scyllaOneShardDialer struct {
-	shardAwareAddress string
-	shardID           int
-	nrShards          int
-	dialer            Dialer
+type scyllaDialer struct {
+	dialer    Dialer
+	logger    StdLogger
+	tlsConfig *tls.Config
+	cfg       *ClusterConfig
 }
 
 const scyllaShardAwarePortFallbackDuration time.Duration = 5 * time.Minute
 
-func (sosd *scyllaOneShardDialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
-	iter := newScyllaPortIterator(sosd.shardID, sosd.nrShards)
+func (sd *scyllaDialer) DialHost(ctx context.Context, host *HostInfo) (*DialedHost, error) {
+	ip := host.ConnectAddress()
+	port := host.Port()
+
+	if !validIpAddr(ip) {
+		return nil, fmt.Errorf("host missing connect ip address: %v", ip)
+	} else if port == 0 {
+		return nil, fmt.Errorf("host missing port: %v", port)
+	}
+
+	addr := host.HostnameAndPort()
+	conn, err := sd.dialer.DialContext(ctx, "tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return sd.wrapTLS(conn, addr)
+}
+
+func (sd *scyllaDialer) DialShard(ctx context.Context, host *HostInfo, shardID, nrShards int) (*DialedHost, error) {
+	ip := host.ConnectAddress()
+	port := host.Port()
+
+	if !validIpAddr(ip) {
+		return nil, fmt.Errorf("host missing connect ip address: %v", ip)
+	} else if port == 0 {
+		return nil, fmt.Errorf("host missing port: %v", port)
+	}
+
+	iter := newScyllaPortIterator(shardID, nrShards)
+
+	addr := host.HostnameAndPort()
+
+	var shardAwarePort uint16
+	if sd.tlsConfig != nil {
+		shardAwarePort = host.ScyllaShardAwarePortTLS()
+	} else {
+		shardAwarePort = host.ScyllaShardAwarePort()
+	}
+
+	var shardAwareAddress string
+	if shardAwarePort != 0 {
+		tIP, tPort := sd.cfg.translateAddressPort(host.UntranslatedConnectAddress(), int(shardAwarePort))
+		shardAwareAddress = net.JoinHostPort(tIP.String(), strconv.Itoa(tPort))
+	}
 
 	if gocqlDebug {
-		Logger.Printf("scylla: connecting to shard %d", sosd.shardID)
+		sd.logger.Printf("scylla: connecting to shard %d", shardID)
+	}
+
+	conn, err := sd.dialShardAware(ctx, addr, shardAwareAddress, iter)
+	if err != nil {
+		return nil, err
+	}
+
+	return sd.wrapTLS(conn, addr)
+}
+
+func (sd *scyllaDialer) wrapTLS(conn net.Conn, addr string) (*DialedHost, error) {
```

Comment:
This function looks very similar to `tlsConfigForAddr`. Would it be possible to deduplicate the logic?

Reply:
`wrapTLS` uses `tlsConfigForAddr`. Did you mean `defaultHostDialer.DialHost`? Anyway, I made wrapTLS a standalone function and used it in defaultHostDialer.DialHost as well.